api/src/test/java/org/pmiops/workbench/api/DataSetControllerTest.java

```diff
@@ -0,0 +1,419 @@
+        WORKSPACE_NAMESPACE, WORKSPACE_NAME, conceptSetTwoRequest).getBody();
+    CONCEPT_SET_TWO_ID = conceptSetTwo.getId();
+
+    when(participantCounter.buildParticipantIdQuery(any())).thenReturn(
+        QueryJobConfiguration.newBuilder("SELECT * FROM person_id from `${projectId}.${dataSetId}.person` person").build());
+    when(bigQueryService.filterBigQueryConfig(any())).thenReturn(
+        QueryJobConfiguration.newBuilder("SELECT * FROM person_id from `all-of-us-ehr-dev.synthetic_cdr20180606.person` person").build());
+  }
+
+
+
+  @Test(expected = BadRequestException.class)
+  public void testGetQueryFailsWithNoCohort() {
+    DataSet dataSet = buildEmptyDataSet();
+    dataSet = dataSet.addConceptSetIdsItem(CONCEPT_SET_ONE_ID);
+
+    dataSetController.getQueryFromDataSet(WORKSPACE_NAMESPACE, WORKSPACE_NAME, dataSet);
+  }
+
+  @Test(expected = BadRequestException.class)
+  public void testGetQueryFailsWithNoConceptSet() {
+    DataSet dataSet = buildEmptyDataSet();
+    dataSet = dataSet.addCohortIdsItem(COHORT_ONE_ID);
+
+    dataSetController.getQueryFromDataSet(WORKSPACE_NAMESPACE, WORKSPACE_NAME, dataSet);
+  }
+
+  @Test
+  public void testGetQueryDropsQueriesWithNoValue() {
+    DataSet dataSet = buildEmptyDataSet();
+    dataSet = dataSet.addCohortIdsItem(COHORT_ONE_ID);
+    dataSet = dataSet.addConceptSetIdsItem(CONCEPT_SET_ONE_ID);
+
+    DataSetQueryResponse response = dataSetController.getQueryFromDataSet(WORKSPACE_NAMESPACE, WORKSPACE_NAME, dataSet).getBody();
+    assertThat(response.getQueryList()).isEmpty();
+  }
+
+  @Test
+  public void testGetQuery() {
+    DataSet dataSet = buildEmptyDataSet();
+    dataSet = dataSet.addCohortIdsItem(COHORT_ONE_ID);
+    dataSet = dataSet.addConceptSetIdsItem(CONCEPT_SET_ONE_ID);
+    List<DomainValue> domainValues = new ArrayList<>();
+    domainValues.add(new DomainValue().value("PERSON_ID"));
+    ValueSet valueSet = new ValueSet().domain(Domain.CONDITION).values(new DomainValuesResponse().items(domainValues));
+    dataSet.addValuesItem(valueSet);
+
+
+    List<String> selectStrings = new ArrayList<>();
+    selectStrings.add("PERSON_ID");
+    List<String> joinStrings = new ArrayList<>();
+    joinStrings.add("FROM `all-of-us-ehr-dev.synthetic_cdr20180606.condition_occurrence` c_occurrence");
+
+    dataSetController = spy(dataSetController);
+    doReturn(new ValuesLinkingPair(selectStrings, joinStrings)).when(dataSetController).getValueSelectsAndJoins(valueSet, Domain.CONDITION);
+
+    DataSetQueryResponse response = dataSetController.getQueryFromDataSet(WORKSPACE_NAMESPACE, WORKSPACE_NAME, dataSet).getBody();
+    assertThat(response.getQueryList().size()).isEqualTo(1);
+    verify(dataSetController, times(1)).getValueSelectsAndJoins(valueSet, Domain.CONDITION);
+    assertThat(response.getQueryList().get(0).getQuery()).matches("SELECT(.*)WHERE CONDITION_CONCEPT_ID IN \\((.*)\\)(.*)PERSON_ID IN \\(SELECT(.*)");
```

Comment:
This seems like a very minimal check against the generated SQL structure. Are we sure this is sufficient to protect against regressions in our generation logic?

I'd also love to see a bit more clarity on what's being verified by this test. If you could break this down into a set of smaller, distinct regexps, it may be clearer which parts of the structure are important to maintain as invariants.

Separately, this test doesn't really help me as a reader understand what the resulting SQL looks like, which is often part of the goal of unit tests. I wouldn't actually be against verifying the exact SQL string that gets generated here (if it's a reasonable length), since the SQL string really is part of our public-facing API.

Reply:
I can definitely try. Let me... actually let me see if I can get an exact SQL string match as part of it. I agree that would be much better.